shoop/admin/modules/orders/static_src/create/reducers/lines.js

```diff
@@ -68,24 +106,60 @@ function setLineProperty(state, {payload}) {
                     updates.sku = null;
                 }
                 if (value === "text") {
+                    updates = getDiscountsAndTotal(0, line.baseUnitPrice, 0);
                     updates.unitPrice = 0;
                     updates.quantity = 0;
                 }
+                updates.type = value;
                 break;
             case "quantity":
-                updates.quantity = Math.max(0, parseFloat(value));
+                var quantity = Math.max(0, parseFloat(value));
+                updates  = getDiscountsAndTotal(quantity, line.baseUnitPrice, line.unitPrice);
+                updates.quantity = quantity;
                 break;
             case "unitPrice":
-                updates.unitPrice = Math.max(0, parseFloat(value));
+                var unitPrice = parseFloat(value);
+                updates  = getDiscountsAndTotal(line.quantity, line.baseUnitPrice, unitPrice);
+                updates.unitPrice = unitPrice;
+                break;
+            case "discountPercent":
+                var discountPercent = Math.max(0, parseFloat(value));
+                updates  = getDiscountsAndTotal(
+                    line.quantity, line.baseUnitPrice, (line.baseUnitPrice * (1 - (discountPercent / 100))), true
+                );
+                updates.discountPercent = discountPercent;
+                break;
+            case "discountAmount":
+                var discountAmount = Math.max(0, parseFloat(value));
+                updates  = getDiscountsAndTotal(
+                    line.quantity, line.baseUnitPrice, (line.baseUnitPrice - (discountAmount / line.quantity)), true
+                );
+                updates.discountAmount = discountAmount;
+                break;
+            case "total":
+                var total = parseFloat(value);
+                updates  = getDiscountsAndTotal(line.quantity, line.baseUnitPrice, (total / line.quantity), true);
+                updates.total = total;
                 break;
         }
     }
     return setLineProperties(state, id, updates);
 }
 
+function updateLines(state, {payload}) {
```

Comment:
You could move this to the top-level reducer in `reducers/index.js`; that way you get direct access to the full state without having to pass the `store` as the payload (which has a terrible code smell).

Then you could change the payload to only have `dispatch` (`{dispatch: store.dispatch.bind(store)}`, and you'd end up with something like

```javascript
function updateLines(state, {payload}) {
    const {dispatch} = payload;
    state.lines.forEach((line) => {
        if (line.product) {
            dispatch(retrieveProductData({id: line.product.id, forLine: line.id, quantity: line.quantity}));
        }
    });
    return state;
}
```